v2: Do not load `.tar.bz2` record if there's a `.conda` artifact counterpart

`defaults` has both `.tar.bz2` and `.conda` artifacts for many of its packages. When loading the repodata for e.g. `defaults/noarch`, `conda`'s `SubdirData` will not list the tar.bz2 artifacts for those packages that feature a .conda copy. This saves a lot of items.

I have a local test I haven't pushed yet, but this is the gist:

```
File "/workspaces/conda-libmamba-solver/tests/test_index2.py", line 84, in test_defaults_use_only_tar_bz2
    assert libmamba_tar_bz2_total + libmamba_dot_conda_total == conda_tar_bz2_total + conda_dot_conda_total
AssertionError: assert (19967 + 14276) == (5691 + 14276)
```

I don't know if you want to do this by default, but maybe it can be a flag exposed in the `Database` signatures?

--- a/libmamba/include/mamba/solver/libsolv/database.hpp
+++ b/libmamba/include/mamba/solver/libsolv/database.hpp
@@ -73,7 +73,7 @@ namespace mamba::solver::libsolv
             std::string_view url,
             const std::string& channel_id,
             PipAsPythonDependency add = PipAsPythonDependency::No,
-            PackageTypes only_tar = PackageTypes::CondaOrElseTarBz2,
+            PackageTypes package_types = PackageTypes::CondaOrElseTarBz2,
             VerifyPackages verify_packages = VerifyPackages::No,
             RepodataParser parser = RepodataParser::Mamba
         ) -> expected_t<RepoInfo>;

--- a/libmamba/include/mamba/solver/libsolv/database.hpp
+++ b/libmamba/include/mamba/solver/libsolv/database.hpp
@@ -73,7 +73,7 @@ namespace mamba::solver::libsolv
             std::string_view url,
             const std::string& channel_id,
             PipAsPythonDependency add = PipAsPythonDependency::No,
-            UseOnlyTarBz2 only_tar = UseOnlyTarBz2::No,
+            PackageTypes only_tar = PackageTypes::CondaOrElseTarBz2,
             VerifyPackages verify_packages = VerifyPackages::No,
             RepodataParser parser = RepodataParser::Mamba
         ) -> expected_t<RepoInfo>;

--- a/libmamba/include/mamba/solver/libsolv/parameters.hpp
+++ b/libmamba/include/mamba/solver/libsolv/parameters.hpp
@@ -25,13 +25,15 @@ namespace mamba::solver::libsolv
         Yes = true,
     };
 
-    enum class UseOnlyTarBz2
+    enum class PackageTypes
     {
-        No = false,
-        Yes = true,
+        CondaOnly,
+        TarBz2Only,
+        CondaAndTarBz2,
+        CondaOrElseTarBz2,
     };
 
-    enum class VerifyPackages
+    enum class VerifyPackages : bool
     {
         No = false,
         Yes = true,

--- a/libmamba/src/core/package_database_loader.cpp
+++ b/libmamba/src/core/package_database_loader.cpp
@@ -93,13 +93,16 @@ namespace mamba
             .and_then(
                 [&](fs::u8path&& repodata_json)
                 {
+                    using PackageTypes = solver::libsolv::PackageTypes;
+
                     LOG_INFO << "Trying to load repo from json file " << repodata_json;
                     return db.add_repo_from_repodata_json(
                         repodata_json,
                         util::rsplit(subdir.metadata().url(), "/", 1).front(),
                         subdir.channel_id(),
                         add_pip,
-                        static_cast<solver::libsolv::PackageTypes>(ctx.use_only_tar_bz2),
+                        ctx.use_only_tar_bz2 ? PackageTypes::TarBz2Only
+                                             : PackageTypes::CondaOrElseTarBz2,
                         static_cast<solver::libsolv::VerifyPackages>(ctx.validation_params.verify_artifacts
                         ),
                         json_parser

--- a/libmamba/src/core/package_database_loader.cpp
+++ b/libmamba/src/core/package_database_loader.cpp
@@ -99,7 +99,7 @@ namespace mamba
                         util::rsplit(subdir.metadata().url(), "/", 1).front(),
                         subdir.channel_id(),
                         add_pip,
-                        static_cast<solver::libsolv::UseOnlyTarBz2>(ctx.use_only_tar_bz2),
+                        static_cast<solver::libsolv::PackageTypes>(ctx.use_only_tar_bz2),
                         static_cast<solver::libsolv::VerifyPackages>(ctx.validation_params.verify_artifacts
                         ),
                         json_parser

--- a/libmamba/src/solver/libsolv/database.cpp
+++ b/libmamba/src/solver/libsolv/database.cpp
@@ -131,7 +131,7 @@ namespace mamba::solver::libsolv
         std::string_view url,
         const std::string& channel_id,
         PipAsPythonDependency add,
-        PackageTypes only_tar,
+        PackageTypes package_types,
         VerifyPackages verify_packages,
         RepodataParser parser
     ) -> expected_t<RepoInfo>
@@ -163,7 +163,7 @@ namespace mamba::solver::libsolv
                     verify_artifacts
                 );
             }
-            return libsolv_read_json(repo, path, use_only_tar_bz2, verify_artifacts)
+            return libsolv_read_json(repo, path, package_types, verify_artifacts)
                 .transform(
                     [&url, &channel_id](solv::ObjRepoView p_repo)
                     {

--- a/libmamba/src/solver/libsolv/database.cpp
+++ b/libmamba/src/solver/libsolv/database.cpp
@@ -131,7 +131,7 @@ namespace mamba::solver::libsolv
         std::string_view url,
         const std::string& channel_id,
         PipAsPythonDependency add,
-        UseOnlyTarBz2 only_tar,
+        PackageTypes only_tar,
         VerifyPackages verify_packages,
         RepodataParser parser
     ) -> expected_t<RepoInfo>

--- a/libmamba/src/solver/libsolv/database.cpp
+++ b/libmamba/src/solver/libsolv/database.cpp
@@ -136,7 +136,6 @@ namespace mamba::solver::libsolv
         RepodataParser parser
     ) -> expected_t<RepoInfo>
     {
-        const auto use_only_tar_bz2 = static_cast<bool>(only_tar);
         const auto verify_artifacts = static_cast<bool>(verify_packages);
 
         if (!fs::exists(path))
@@ -159,7 +158,7 @@ namespace mamba::solver::libsolv
                     path,
                     std::string(url),
                     channel_id,
-                    use_only_tar_bz2,
+                    package_types,
                     verify_artifacts
                 );
             }

--- a/libmamba/src/solver/libsolv/helpers.cpp
+++ b/libmamba/src/solver/libsolv/helpers.cpp
@@ -401,35 +401,6 @@ namespace mamba::solver::libsolv
                 }
             }
         }
-
-        void set_repo_solvables_with_sigs(
-            solv::ObjPool& pool,
-            solv::ObjRepoView repo,
-            const specs::CondaURL& parsed_url,
-            const std::string& channel_id,
-            const std::string& default_subdir,
-            const simdjson::dom::object& packages,
-            const simdjson::dom::object& repodata,
-            bool verify_artifacts
-        )
-        {
-            auto signatures = std::optional<std::reference_wrapper<const simdjson::dom::object>>();
-            if (auto maybe_sigs = repodata["signatures"].get_object();
-                !maybe_sigs.error() && verify_artifacts)
-            {
-                signatures = maybe_sigs.value();
-            }
-
-            set_repo_solvables(  //
-                pool,
-                repo,
-                parsed_url,
-                channel_id,
-                default_subdir,
-                packages,
-                signatures
-            );
-        }
     }
 
     auto libsolv_read_json(
@@ -499,31 +470,36 @@ namespace mamba::solver::libsolv
                                     .or_else([](specs::ParseError&& err) { throw std::move(err); })
                                     .value();
 
+        auto signatures = std::optional<std::reference_wrapper<const simdjson::dom::object>>();
+        if (auto maybe_sigs = repodata["signatures"].get_object();
+            !maybe_sigs.error() && verify_artifacts)
+        {
+            signatures = maybe_sigs.value();
+        }
+
         if (auto pkgs = repodata["packages"].get_object(); !pkgs.error())
         {
-            set_repo_solvables_with_sigs(
+            set_repo_solvables(  //
                 pool,
                 repo,
-                /*repo_url,*/ parsed_url,
+                parsed_url,
                 channel_id,
                 default_subdir,
                 pkgs.value(),
-                repodata,
-                verify_artifacts
+                signatures
             );
         }
 
         if (auto pkgs = repodata["packages.conda"].get_object(); !pkgs.error() && !only_tar_bz2)
         {
-            set_repo_solvables_with_sigs(
+            set_repo_solvables(  //
                 pool,
                 repo,
-                /*repo_url,*/ parsed_url,
+                parsed_url,
                 channel_id,
                 default_subdir,
                 pkgs.value(),
-                repodata,
-                verify_artifacts
+                signatures
             );
         }
 

--- a/libmamba/src/solver/libsolv/helpers.cpp
+++ b/libmamba/src/solver/libsolv/helpers.cpp
@@ -20,6 +20,7 @@
 
 #include "mamba/core/output.hpp"
 #include "mamba/core/util.hpp"
+#include "mamba/specs/archive.hpp"
 #include "mamba/specs/conda_url.hpp"
 #include "mamba/util/cfile.hpp"
 #include "mamba/util/random.hpp"
@@ -362,10 +363,53 @@ namespace mamba::solver::libsolv
             return true;
         }
 
+        template <typename Filter, typename OnParsed>
+        void set_repo_solvables_impl(
+            solv::ObjPool& pool,
+            solv::ObjRepoView repo,
+            const specs::CondaURL& repo_url,
+            const std::string& channel_id,
+            const std::string& default_subdir,
+            const simdjson::dom::object& packages,
+            const std::optional<simdjson::dom::object>& signatures,
+            Filter&& filter,
+            OnParsed&& on_parsed
+        )
+        {
+            std::string filename = {};
+            for (const auto& [fn, pkg] : packages)
+            {
+                if (filter(fn))
+                {
+                    auto [id, solv] = repo.add_solvable();
+                    filename = fn;
+                    const bool parsed = set_solvable(
+                        pool,
+                        repo_url,
+                        channel_id,
+                        solv,
+                        filename,
+                        pkg,
+                        signatures,
+                        default_subdir
+                    );
+                    if (parsed)
+                    {
+                        on_parsed(fn);
+                        LOG_DEBUG << "Adding package record to repo " << fn;
+                    }
+                    else
+                    {
+                        repo.remove_solvable(id, /* reuse_id= */ true);
+                        LOG_WARNING << "Failed to parse from repodata " << fn;
+                    }
+                }
+            }
+        }
+
         void set_repo_solvables(
             solv::ObjPool& pool,
             solv::ObjRepoView repo,
-            // const std::string& repo_url_str,
             const specs::CondaURL& repo_url,
             const std::string& channel_id,
             const std::string& default_subdir,
@@ -373,32 +417,69 @@ namespace mamba::solver::libsolv
             const std::optional<simdjson::dom::object>& signatures
         )
         {
-            std::string filename = {};
-            for (const auto& [fn, pkg] : packages)
-            {
-                auto [id, solv] = repo.add_solvable();
-                filename = fn;
-                const bool parsed = set_solvable(
-                    pool,
-                    // repo_url_str,
-                    repo_url,
-                    channel_id,
-                    solv,
-                    filename,
-                    pkg,
-                    signatures,
-                    default_subdir
-                );
-                if (parsed)
-                {
-                    LOG_DEBUG << "Adding package record to repo " << fn;
-                }
-                else
-                {
-                    repo.remove_solvable(id, /* reuse_id= */ true);
-                    LOG_WARNING << "Failed to parse from repodata " << fn;
-                }
-            }
+            return set_repo_solvables_impl(
+                pool,
+                repo,
+                repo_url,
+                channel_id,
+                default_subdir,
+                packages,
+                signatures,
+                /* filter= */ [](const auto&) { return true; },
+                /* on_parsed= */ [](const auto&) {}
+            );
+        }
+
+        auto set_repo_solvables_and_return_added_filename_stem(
+            solv::ObjPool& pool,
+            solv::ObjRepoView repo,
+            const specs::CondaURL& repo_url,
+            const std::string& channel_id,
+            const std::string& default_subdir,
+            const simdjson::dom::object& packages,
+            const std::optional<simdjson::dom::object>& signatures
+        ) -> util::flat_set<std::string_view>
+        {
+            auto filenames = std::vector<std::string_view>();
+            set_repo_solvables_impl(
+                pool,
+                repo,
+                repo_url,
+                channel_id,
+                default_subdir,
+                packages,
+                signatures,
+                /* filter= */ [](const auto&) { return true; },
+                /* on_parsed= */ [&](const auto& fn)
+                { filenames.push_back(specs::strip_archive_extension(fn)); }
+            );
+            // Sort only once
+            return util::flat_set<std::string_view>{ std::move(filenames) };
+        }
+
+        void set_repo_solvables_if_not_already_set(
+            solv::ObjPool& pool,
+            solv::ObjRepoView repo,
+            const specs::CondaURL& repo_url,
+            const std::string& channel_id,
+            const std::string& default_subdir,
+            const simdjson::dom::object& packages,
+            const std::optional<simdjson::dom::object>& signatures,
+            const util::flat_set<std::string_view>& added
+        )
+        {
+            return set_repo_solvables_impl(
+                pool,
+                repo,
+                repo_url,
+                channel_id,
+                default_subdir,
+                packages,
+                signatures,
+                /* filter= */ [&](const auto& fn)
+                { return !added.contains(specs::strip_archive_extension(fn)); },
+                /* on_parsed= */ [&](const auto&) {}
+            );
         }
     }
 
@@ -456,7 +537,7 @@ namespace mamba::solver::libsolv
         const fs::u8path& filename,
         const std::string& repo_url,
         const std::string& channel_id,
-        bool only_tar_bz2,
+        PackageTypes package_types,
         bool verify_artifacts
     ) -> expected_t<solv::ObjRepoView>
     {
@@ -485,30 +566,64 @@ namespace mamba::solver::libsolv
             signatures = std::move(maybe_sigs).value();
         }
 
-        if (auto pkgs = repodata["packages"].get_object(); !pkgs.error())
+        if (package_types == PackageTypes::CondaOrElseTarBz2)
         {
-            set_repo_solvables(  //
-                pool,
-                repo,
-                parsed_url,
-                channel_id,
-                default_subdir,
-                pkgs.value(),
-                signatures
-            );
+            auto added = util::flat_set<std::string_view>();
+            if (auto pkgs = repodata["packages.conda"].get_object(); !pkgs.error())
+            {
+                added = set_repo_solvables_and_return_added_filename_stem(  //
+                    pool,
+                    repo,
+                    parsed_url,
+                    channel_id,
+                    default_subdir,
+                    pkgs.value(),
+                    signatures
+                );
+            }
+            if (auto pkgs = repodata["packages"].get_object(); !pkgs.error())
+            {
+                set_repo_solvables_if_not_already_set(  //
+                    pool,
+                    repo,
+                    parsed_url,
+                    channel_id,
+                    default_subdir,
+                    pkgs.value(),
+                    signatures,
+                    added
+                );
+            }
         }
-
-        if (auto pkgs = repodata["packages.conda"].get_object(); !pkgs.error() && !only_tar_bz2)
+        else
         {
-            set_repo_solvables(  //
-                pool,
-                repo,
-                parsed_url,
-                channel_id,
-                default_subdir,
-                pkgs.value(),
-                signatures
-            );
+            if (auto pkgs = repodata["packages"].get_object();
+                !pkgs.error() && (package_types != PackageTypes::CondaOnly))
+            {
+                set_repo_solvables(  //
+                    pool,
+                    repo,
+                    parsed_url,
+                    channel_id,
+                    default_subdir,
+                    pkgs.value(),
+                    signatures
+                );
+            }
+
+            if (auto pkgs = repodata["packages.conda"].get_object();
+                !pkgs.error() && (package_types != PackageTypes::TarBz2Only))
+            {
+                set_repo_solvables(  //
+                    pool,
+                    repo,
+                    parsed_url,
+                    channel_id,
+                    default_subdir,
+                    pkgs.value(),
+                    signatures
+                );
+            }
         }
 
         return { repo };

--- a/libmamba/src/solver/libsolv/helpers.cpp
+++ b/libmamba/src/solver/libsolv/helpers.cpp
@@ -405,14 +405,23 @@ namespace mamba::solver::libsolv
     auto libsolv_read_json(
         solv::ObjRepoView repo,
         const fs::u8path& filename,
-        bool only_tar_bz2,
+        PackageTypes types,
         bool verify_artifacts
     ) -> expected_t<solv::ObjRepoView>
     {
+        if ((types != PackageTypes::TarBz2Only) && (types != PackageTypes::CondaOrElseTarBz2))
+        {
+            return make_unexpected(
+                "Invalid PackageTypes option for libsolv repodata.json parser:"
+                " supported types are TarBz2Only and CondaOrElseTarBz2.",
+                mamba_error_code::repodata_not_loaded
+            );
+        }
+
         LOG_INFO << "Reading repodata.json file " << filename << " for repo " << repo.name()
                  << " using libsolv";
 
-        int flags = only_tar_bz2 ? CONDA_ADD_USE_ONLY_TAR_BZ2 : 0;
+        int flags = (types == PackageTypes::TarBz2Only) ? CONDA_ADD_USE_ONLY_TAR_BZ2 : 0;
         if (verify_artifacts)
         {
             // cf.

--- a/libmamba/src/solver/libsolv/helpers.cpp
+++ b/libmamba/src/solver/libsolv/helpers.cpp
@@ -4,7 +4,6 @@
 //
 // The full license is in the file LICENSE, distributed with this software.
 
-#include <functional>
 #include <string>
 #include <string_view>
 #include <type_traits>
@@ -155,7 +154,7 @@ namespace mamba::solver::libsolv
         void set_solv_signatures(
             solv::ObjSolvableView solv,
             const std::string& filename,
-            std::optional<std::reference_wrapper<const simdjson::dom::object>> signatures
+            const std::optional<simdjson::dom::object>& signatures
         )
         {
             // NOTE We need to use an intermediate nlohmann::json object to store signatures
@@ -164,7 +163,7 @@ namespace mamba::solver::libsolv
             nlohmann::json glob_sigs, nested_sigs;
             if (signatures)
             {
-                if (auto sigs = signatures->get()[filename].get_object(); !sigs.error())
+                if (auto sigs = signatures.value()[filename].get_object(); !sigs.error())
                 {
                     for (auto dict : sigs)
                     {
@@ -195,7 +194,7 @@ namespace mamba::solver::libsolv
             solv::ObjSolvableView solv,
             const std::string& filename,
             const simdjson::dom::element& pkg,
-            std::optional<std::reference_wrapper<const simdjson::dom::object>> signatures,
+            const std::optional<simdjson::dom::object>& signatures,
             const std::string& default_subdir
         ) -> bool
         {
@@ -371,7 +370,7 @@ namespace mamba::solver::libsolv
             const std::string& channel_id,
             const std::string& default_subdir,
             const simdjson::dom::object& packages,
-            std::optional<std::reference_wrapper<const simdjson::dom::object>> signatures
+            const std::optional<simdjson::dom::object>& signatures
         )
         {
             std::string filename = {};
@@ -470,11 +469,11 @@ namespace mamba::solver::libsolv
                                     .or_else([](specs::ParseError&& err) { throw std::move(err); })
                                     .value();
 
-        auto signatures = std::optional<std::reference_wrapper<const simdjson::dom::object>>();
+        auto signatures = std::optional<simdjson::dom::object>(std::nullopt);
         if (auto maybe_sigs = repodata["signatures"].get_object();
             !maybe_sigs.error() && verify_artifacts)
         {
-            signatures = maybe_sigs.value();
+            signatures = std::move(maybe_sigs).value();
         }
 
         if (auto pkgs = repodata["packages"].get_object(); !pkgs.error())

--- a/libmamba/src/solver/libsolv/helpers.cpp
+++ b/libmamba/src/solver/libsolv/helpers.cpp
@@ -4,6 +4,7 @@
 //
 // The full license is in the file LICENSE, distributed with this software.
 
+#include <functional>
 #include <string>
 #include <string_view>
 #include <type_traits>
@@ -154,7 +155,7 @@ namespace mamba::solver::libsolv
         void set_solv_signatures(
             solv::ObjSolvableView solv,
             const std::string& filename,
-            const std::optional<simdjson::dom::object>& signatures
+            std::optional<std::reference_wrapper<const simdjson::dom::object>> signatures
         )
         {
             // NOTE We need to use an intermediate nlohmann::json object to store signatures
@@ -163,7 +164,7 @@ namespace mamba::solver::libsolv
             nlohmann::json glob_sigs, nested_sigs;
             if (signatures)
             {
-                if (auto sigs = signatures.value()[filename].get_object(); !sigs.error())
+                if (auto sigs = signatures->get()[filename].get_object(); !sigs.error())
                 {
                     for (auto dict : sigs)
                     {
@@ -194,7 +195,7 @@ namespace mamba::solver::libsolv
             solv::ObjSolvableView solv,
             const std::string& filename,
             const simdjson::dom::element& pkg,
-            const std::optional<simdjson::dom::object>& signatures,
+            std::optional<std::reference_wrapper<const simdjson::dom::object>> signatures,
             const std::string& default_subdir
         ) -> bool
         {
@@ -370,7 +371,7 @@ namespace mamba::solver::libsolv
             const std::string& channel_id,
             const std::string& default_subdir,
             const simdjson::dom::object& packages,
-            const std::optional<simdjson::dom::object>& signatures
+            std::optional<std::reference_wrapper<const simdjson::dom::object>> signatures
         )
         {
             std::string filename = {};
@@ -412,23 +413,22 @@ namespace mamba::solver::libsolv
             bool verify_artifacts
         )
         {
-            if (auto signatures = repodata["signatures"].get_object();
-                !signatures.error() && verify_artifacts)
+            auto signatures = std::optional<std::reference_wrapper<const simdjson::dom::object>>();
+            if (auto maybe_sigs = repodata["signatures"].get_object();
+                !maybe_sigs.error() && verify_artifacts)
             {
-                set_repo_solvables(
-                    pool,
-                    repo,
-                    parsed_url,
-                    channel_id,
-                    default_subdir,
-                    packages,
-                    signatures.value()
-                );
+                signatures = maybe_sigs.value();
             }
-            else
-            {
-                set_repo_solvables(pool, repo, parsed_url, channel_id, default_subdir, packages, std::nullopt);
-            }
+
+            set_repo_solvables(  //
+                pool,
+                repo,
+                parsed_url,
+                channel_id,
+                default_subdir,
+                packages,
+                signatures
+            );
         }
     }
 

--- a/libmamba/src/solver/libsolv/helpers.hpp
+++ b/libmamba/src/solver/libsolv/helpers.hpp
@@ -44,7 +44,7 @@ namespace mamba::solver::libsolv
     [[nodiscard]] auto libsolv_read_json(  //
         solv::ObjRepoView repo,
         const fs::u8path& filename,
-        bool only_tar_bz2,
+        PackageTypes types,
         bool verify_artifacts
     ) -> expected_t<solv::ObjRepoView>;
 

--- a/libmamba/src/solver/libsolv/helpers.hpp
+++ b/libmamba/src/solver/libsolv/helpers.hpp
@@ -54,7 +54,7 @@ namespace mamba::solver::libsolv
         const fs::u8path& filename,
         const std::string& repo_url,
         const std::string& channel_id,
-        bool only_tar_bz2,
+        PackageTypes types,
         bool verify_artifacts
     ) -> expected_t<solv::ObjRepoView>;
 

--- a/libmamba/src/solver/problems_graph.cpp
+++ b/libmamba/src/solver/problems_graph.cpp
@@ -105,7 +105,6 @@ namespace mamba::solver
                             assert(graph.has_node(id_child));
                             assert(graph.has_node(c_parent));
                             conflicts.add(id_child, c_parent);
-                            graph.remove_edge(c_parent, c);
                         }
                         conflicts.remove(id, c);
                         if (!conflicts.has_conflict(c))

--- a/libmamba/tests/data/repodata/conda-forge-numpy-linux-64.json
+++ b/libmamba/tests/data/repodata/conda-forge-numpy-linux-64.json
@@ -213,6 +213,26 @@
       "url": "https://conda.anaconda.org/conda-forge/linux-64/libexpat-2.5.0-hcb278e6_1.conda",
       "version": "2.5.0"
     },
+    "libffi-3.4.2-h7f98852_5.conda": {
+      "build": "h7f98852_5",
+      "build_number": 5,
+      "build_string": "h7f98852_5",
+      "constrains": null,
+      "depends": [
+        "libgcc-ng >=9.4.0"
+      ],
+      "fn": "libffi-3.4.2-h7f98852_5.conda",
+      "license": "MIT",
+      "md5": "d645c6d2ac96843a2bfaccd2d62b3ac3",
+      "name": "libffi",
+      "sha256": "0000000000000000000000000000000000000000000000000000000000000000",
+      "size": 58292,
+      "subdir": "linux-64",
+      "timestamp": 1636488182,
+      "track_features": "",
+      "url": "https://conda.anaconda.org/conda-forge/linux-64/libffi-3.4.2-h7f98852_5.conda",
+      "version": "3.4.2"
+    },
     "libgcc-ng-13.2.0-h807b86a_5.conda": {
       "build": "h807b86a_5",
       "build_number": 5,

--- a/libmamba/tests/src/solver/libsolv/test_database.cpp
+++ b/libmamba/tests/src/solver/libsolv/test_database.cpp
@@ -246,7 +246,7 @@ TEST_SUITE("solver::libsolv::database")
                 "https://conda.anaconda.org/conda-forge/linux-64",
                 "conda-forge",
                 libsolv::PipAsPythonDependency::No,
-                libsolv::UseOnlyTarBz2::Yes
+                libsolv::PackageTypes::CondaOnly
             );
             REQUIRE(repo1.has_value());
             CHECK_EQ(repo1->package_count(), 4);
@@ -263,7 +263,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::UseOnlyTarBz2::No,
+                    libsolv::PackageTypes::CondaAndTarBz2,
                     libsolv::VerifyPackages::Yes,
                     libsolv::RepodataParser::Mamba
                 );
@@ -303,7 +303,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::UseOnlyTarBz2::No,
+                    libsolv::PackageTypes::CondaAndTarBz2,
                     libsolv::VerifyPackages::Yes,
                     libsolv::RepodataParser::Libsolv
                 );
@@ -352,7 +352,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::UseOnlyTarBz2::No,
+                    libsolv::PackageTypes::CondaAndTarBz2,
                     libsolv::VerifyPackages::No,
                     libsolv::RepodataParser::Mamba
                 );
@@ -372,7 +372,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::UseOnlyTarBz2::No,
+                    libsolv::PackageTypes::CondaAndTarBz2,
                     libsolv::VerifyPackages::No,
                     libsolv::RepodataParser::Libsolv
                 );

--- a/libmamba/tests/src/solver/libsolv/test_database.cpp
+++ b/libmamba/tests/src/solver/libsolv/test_database.cpp
@@ -246,10 +246,55 @@ TEST_SUITE("solver::libsolv::database")
                 "https://conda.anaconda.org/conda-forge/linux-64",
                 "conda-forge",
                 libsolv::PipAsPythonDependency::No,
-                libsolv::PackageTypes::CondaOnly
+                libsolv::PackageTypes::TarBz2Only
             );
             REQUIRE(repo1.has_value());
             CHECK_EQ(repo1->package_count(), 4);
+        }
+
+        SUBCASE("Add repo from repodata only .conda")
+        {
+            const auto repodata = mambatests::test_data_dir
+                                  / "repodata/conda-forge-numpy-linux-64.json";
+            auto repo1 = db.add_repo_from_repodata_json(
+                repodata,
+                "https://conda.anaconda.org/conda-forge/linux-64",
+                "conda-forge",
+                libsolv::PipAsPythonDependency::No,
+                libsolv::PackageTypes::CondaOnly
+            );
+            REQUIRE(repo1.has_value());
+            CHECK_EQ(repo1->package_count(), 30);
+        }
+
+        SUBCASE("Add repo from repodata .conda and .tar.bz2")
+        {
+            const auto repodata = mambatests::test_data_dir
+                                  / "repodata/conda-forge-numpy-linux-64.json";
+            auto repo1 = db.add_repo_from_repodata_json(
+                repodata,
+                "https://conda.anaconda.org/conda-forge/linux-64",
+                "conda-forge",
+                libsolv::PipAsPythonDependency::No,
+                libsolv::PackageTypes::CondaAndTarBz2
+            );
+            REQUIRE(repo1.has_value());
+            CHECK_EQ(repo1->package_count(), 34);
+        }
+
+        SUBCASE("Add repo from repodata .conda or else .tar.bz2")
+        {
+            const auto repodata = mambatests::test_data_dir
+                                  / "repodata/conda-forge-numpy-linux-64.json";
+            auto repo1 = db.add_repo_from_repodata_json(
+                repodata,
+                "https://conda.anaconda.org/conda-forge/linux-64",
+                "conda-forge",
+                libsolv::PipAsPythonDependency::No,
+                libsolv::PackageTypes::CondaOrElseTarBz2
+            );
+            REQUIRE(repo1.has_value());
+            CHECK_EQ(repo1->package_count(), 33);
         }
 
         SUBCASE("Add repo from repodata with verifying packages signatures")
@@ -263,7 +308,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::PackageTypes::CondaAndTarBz2,
+                    libsolv::PackageTypes::CondaOrElseTarBz2,
                     libsolv::VerifyPackages::Yes,
                     libsolv::RepodataParser::Mamba
                 );
@@ -303,7 +348,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::PackageTypes::CondaAndTarBz2,
+                    libsolv::PackageTypes::CondaOrElseTarBz2,
                     libsolv::VerifyPackages::Yes,
                     libsolv::RepodataParser::Libsolv
                 );
@@ -352,7 +397,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::PackageTypes::CondaAndTarBz2,
+                    libsolv::PackageTypes::CondaOrElseTarBz2,
                     libsolv::VerifyPackages::No,
                     libsolv::RepodataParser::Mamba
                 );
@@ -372,7 +417,7 @@ TEST_SUITE("solver::libsolv::database")
                     "https://conda.anaconda.org/conda-forge/linux-64",
                     "conda-forge",
                     libsolv::PipAsPythonDependency::No,
-                    libsolv::PackageTypes::CondaAndTarBz2,
+                    libsolv::PackageTypes::CondaOrElseTarBz2,
                     libsolv::VerifyPackages::No,
                     libsolv::RepodataParser::Libsolv
                 );

--- a/libmambapy/src/libmambapy/bindings/solver_libsolv.cpp
+++ b/libmambapy/src/libmambapy/bindings/solver_libsolv.cpp
@@ -39,11 +39,13 @@ namespace mambapy
             .def(py::init([](bool val) { return static_cast<PipAsPythonDependency>(val); }));
         py::implicitly_convertible<py::bool_, PipAsPythonDependency>();
 
-        py::enum_<UseOnlyTarBz2>(m, "UseOnlyTarBz2")
-            .value("No", UseOnlyTarBz2::No)
-            .value("Yes", UseOnlyTarBz2::Yes)
-            .def(py::init([](bool val) { return static_cast<UseOnlyTarBz2>(val); }));
-        py::implicitly_convertible<py::bool_, UseOnlyTarBz2>();
+        py::enum_<PackageTypes>(m, "PackageTypes")
+            .value("CondaOnly", PackageTypes::CondaOnly)
+            .value("TarBz2Only", PackageTypes::TarBz2Only)
+            .value("CondaAndTarBz2", PackageTypes::CondaAndTarBz2)
+            .value("CondaOrElseTarBz2", PackageTypes::CondaOrElseTarBz2)
+            .def(py::init(&enum_from_str<PackageTypes>));
+        py::implicitly_convertible<py::str, PackageTypes>();
 
         py::enum_<VerifyPackages>(m, "VerifyPackages")
             .value("No", VerifyPackages::No)
@@ -124,7 +126,7 @@ namespace mambapy
                 py::arg("url"),
                 py::arg("channel_id"),
                 py::arg("add_pip_as_python_dependency") = PipAsPythonDependency::No,
-                py::arg("use_only_tar_bz2") = UseOnlyTarBz2::No,
+                py::arg("package_types") = PackageTypes::CondaOrElseTarBz2,
                 py::arg("verify_packages") = VerifyPackages::No,
                 py::arg("repodata_parser") = RepodataParser::Mamba
             )

--- a/libmambapy/tests/test_solver_libsolv.py
+++ b/libmambapy/tests/test_solver_libsolv.py
@@ -189,7 +189,7 @@ def tmp_repodata_json(tmp_path):
     ["add_pip_as_python_dependency", "package_types", "repodata_parser"],
     itertools.product(
         [True, False],
-        ["CondaOnly", "TarBz2Only", "CondaAndTarBz2", "CondaOrElseTarBz2"],
+        ["TarBz2Only", "CondaOrElseTarBz2"],
         ["Mamba", "Libsolv"],
     ),
 )
@@ -217,8 +217,8 @@ def test_Database_RepoInfo_from_repodata(
 
     pkgs = db.packages_in_repo(repo)
     assert len(pkgs) == repo.package_count()
-    assert pkgs[0].name == "python"
-    assert pkgs[0].dependencies == [] if add_pip_as_python_dependency else ["pip"]
+    python_pkg = next(p for p in pkgs if p.name == "python")
+    assert python_pkg.dependencies == [] if add_pip_as_python_dependency else ["pip"]
 
     # Native serialize repo
     solv_file = tmp_path / "repodata.solv"

--- a/libmambapy/tests/test_solver_libsolv.py
+++ b/libmambapy/tests/test_solver_libsolv.py
@@ -39,11 +39,16 @@ def test_PipASPythonDependency():
     assert libsolv.PipAsPythonDependency(True) == libsolv.PipAsPythonDependency.Yes
 
 
-def test_UseOnlyTarBz2():
-    assert libsolv.UseOnlyTarBz2.No.name == "No"
-    assert libsolv.UseOnlyTarBz2.Yes.name == "Yes"
+def test_PackageTypes():
+    assert libsolv.PackageTypes.CondaOnly.name == "CondaOnly"
+    assert libsolv.PackageTypes.TarBz2Only.name == "TarBz2Only"
+    assert libsolv.PackageTypes.CondaAndTarBz2.name == "CondaAndTarBz2"
+    assert libsolv.PackageTypes.CondaOrElseTarBz2.name == "CondaOrElseTarBz2"
 
-    assert libsolv.UseOnlyTarBz2(True) == libsolv.UseOnlyTarBz2.Yes
+    assert libsolv.PackageTypes("TarBz2Only") == libsolv.PackageTypes.TarBz2Only
+
+    with pytest.raises(KeyError):
+        libsolv.RepodataParser("tarbz2-only")
 
 
 def test_VerifyPackages():
@@ -181,11 +186,15 @@ def tmp_repodata_json(tmp_path):
 
 
 @pytest.mark.parametrize(
-    ["add_pip_as_python_dependency", "use_only_tar_bz2", "repodata_parser"],
-    itertools.product([True, False], [True, False], ["Mamba", "Libsolv"]),
+    ["add_pip_as_python_dependency", "package_types", "repodata_parser"],
+    itertools.product(
+        [True, False],
+        ["CondaOnly", "TarBz2Only", "CondaAndTarBz2", "CondaOrElseTarBz2"],
+        ["Mamba", "Libsolv"],
+    ),
 )
 def test_Database_RepoInfo_from_repodata(
-    tmp_path, tmp_repodata_json, add_pip_as_python_dependency, use_only_tar_bz2, repodata_parser
+    tmp_path, tmp_repodata_json, add_pip_as_python_dependency, package_types, repodata_parser
 ):
     db = libsolv.Database(libmambapy.specs.ChannelResolveParams())
 
@@ -198,12 +207,12 @@ def test_Database_RepoInfo_from_repodata(
         url=url,
         channel_id=channel_id,
         add_pip_as_python_dependency=add_pip_as_python_dependency,
-        use_only_tar_bz2=use_only_tar_bz2,
+        package_types=package_types,
         repodata_parser=repodata_parser,
     )
     db.set_installed_repo(repo)
 
-    assert repo.package_count() == 1 if use_only_tar_bz2 else 2
+    assert repo.package_count() == 1 if package_types in ["TarBz2Only", "CondaOnly"] else 2
     assert db.package_count() == repo.package_count()
 
     pkgs = db.packages_in_repo(repo)
@@ -228,7 +237,7 @@ def test_Database_RepoInfo_from_repodata(
         channel_id=channel_id,
         add_pip_as_python_dependency=add_pip_as_python_dependency,
     )
-    assert repo_loaded.package_count() == 1 if use_only_tar_bz2 else 2
+    assert repo_loaded.package_count() == 1 if package_types in ["TarBz2Only", "CondaOnly"] else 2
 
 
 def test_Database_RepoInfo_from_repodata_error():

